### PR TITLE
[BOOKINGSG-9265][GZ] improve theme breakpoint utils

### PR DESCRIPTION
**Pull request by ghazwan-gt**
**Type of changes**

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [X] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing apis or functionality to change)

**Description of changes**

[Improve breakpoint token utilities](https://sgtechstack.atlassian.net/browse/BOOKINGSG-9265)
-  Improve the breakpoint hooks to be SSR-safe
- Add storybook documentation on how to use the hooks
- Add e2e test to assert the hooks works on SSR
- Combine the previous media query specificity e2e test to be with the new test
- Update some import path on the theme to prevent circular dependency 

**Additional Info**
The next follow-up to this changes is to:
- Update all `react-responsive` & `useSafeMaxWidthMediaQuery` usage to the new hooks
- Remove `react-responsive` package
- Delete `useSafeMaxWidthMediaQuery` utils

**Checklist**

-   [X] Changes follow the project guidelines in [CONTRIBUTING.md](https://github.com/LifeSG/react-design-system/blob/master/CONTRIBUTING.md) and [CONVENTIONS.md](https://github.com/LifeSG/react-design-system/blob/master/CONVENTIONS.md)
-   [ ] Looks good on mobile and tablet
-   [X] Updated documentation
-   [X] Added/updated tests

### CODE REVIEW

**Comment by qroll** [comment]
landscape MediaQuery tokens for use in styles - will they be included in this PR?

**Comment by ghazwan-gt** [comment]
> landscape MediaQuery tokens for use in styles - will they be included in this PR?

Oh right, forgot to add this one. On it

**Comment by qroll** [comment]
> > landscape MediaQuery tokens for use in styles - will they be included in this PR?
> 
> Oh right, forgot to add this one. On it

actually, it may not be needed after we changed how breakpoint detection works

ESignature needed `(orientation: landscape) and (max-height: 480px)` because on mobile devices, the device height is still treated as the screen height rather than the screen width when rotated to landscape

but in theme provider we are now checking against `window.innerWidth` - I do not know if that rotates accordingly

File: src/theme/utils/use-media-query.ts [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        minWidth: `${breakpoint}-min`,
+        orientation,
+    });
+
+export const useMaxWidthMediaQuery = (
+    breakpoint: BreakpointName,
+    orientation?: MediaQueryOrientation
+) =>
+    useMediaQuery({
+        maxWidth: `${breakpoint}-max`,
```
Comment (qroll) [inline src/theme/utils/use-media-query.ts]: there's no xxl max, so we need to restrict the typing here

File: src/theme/utils/use-media-query.ts [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    const resolvedMaxWidth = useResolveBreakpointToken(
+        getWidthCandidate(maxWidth)
+    );
+    const normalizedMinWidth = normalizeCssLengthValue(resolvedMinWidth);
+    const normalizedMaxWidth = normalizeCssLengthValue(resolvedMaxWidth);
+    const clauses = [
+        getMediaQueryClause("min-width", normalizedMinWidth),
+        getMediaQueryClause("max-width", normalizedMaxWidth),
+        orientation ? `(orientation: ${orientation})` : undefined,
+    ].filter((value): value is string => Boolean(value));
```
Comment (qroll) [inline src/theme/utils/use-media-query.ts]:
```suggestion
    ].filter((value): value is string => !!value);
```

File: src/theme/tokens/media-query.ts
Comment (qroll) [inline src/theme/tokens/media-query.ts:1]: the landscape media query token was originally intended for esignature's custom media query, so we don't actually need this

File: src/theme/utils/use-resolve-breakpoint-token.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        : undefined;
+    const resolvedTokenValue = useResolvedTokenValue<
+        BreakpointCSSVariableString,
+        string
+    >({
+        value: cssVariableToken,
+        fallback: fallbackLength ?? breakpointOrLength,
+        isToken: isBreakpointToken,
+        normalizeCustom: (value) => value,
+    });
```
Comment (qroll) [inline src/theme/utils/use-resolve-breakpoint-token.ts]: actually why can't we use `useResolvedTokenValue` directly? that was done previously for the dashed border utils to handle radius/width so this seems pretty similar

```
useResolvedTokenValue({
    value: breakpointOrLength,
    fallback: fallbackLength ?? breakpointOrLength,
    isToken: isBreakpointToken,
    normalizeCustom: (value) => value,
})
```

File: stories/theme/breakpoint/a-breakpoint-introduction.mdx [hunk 4]
```diff
@@ -21,6 +21,7 @@ and **media query** tokens. For grid-based page layouts, **column number** and
 ## Collections
 
 [Default](/docs/foundations-breakpoint-default--docs)
+[Media Query](/docs/foundations-breakpoint-media-query--docs)
```
Comment (qroll) [inline stories/theme/breakpoint/a-breakpoint-introduction.mdx]: collections refer to the theme-specific sets, merge the docs into this file under the `Usage` section instead

we can split this page into two sub-sections: CSS-based style declarations and runtime utilities

File: src/theme/utils/use-media-query.ts [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    const queryString = clauses.join(" and ");
+    const hasMinWidth = Boolean(normalizedMinWidth);
+    const hasMaxWidth = Boolean(normalizedMaxWidth);
+    const defaultMatch = getDefaultMatch(hasMinWidth, hasMaxWidth);
+    const [matches, setMatches] = useState(() =>
+        getCurrentMatch(queryString, defaultMatch)
+    );
+
+    useEffect(() => {
+        setMatches(getCurrentMatch(queryString, defaultMatch));
```
Comment (qroll) [inline src/theme/utils/use-media-query.ts]: we're calling `setMatches` twice in the main effect, could we only limit to 1?

File: e2e/tests/components/theme/__screenshots__/chromium/Theme-media-query-Specificity-Enable-override-media-query-specificity--layout-container-flex-column-specificity.png
Comment (qroll) [inline e2e/tests/components/theme/__screenshots__/chromium/Theme-media-query-Specificity-Enable-override-media-query-specificity--layout-container-flex-column-specificity.png:1]: since the tests have been moved to a different file, remove the previous screenshots generated under `theme.e2e.spec.ts`

File: e2e/nextjs-app/src/app/components/theme/media-query.e2e.tsx [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    useOrientation,
+} from "@lifesg/react-design-system/theme";
+import { useEffect, useState } from "react";
+
+const boolToText = (value: boolean) => String(value);
+
+export default function Story() {
+    const [dimensions, setDimensions] = useState("0 x 0");
+
+    const minXs = useMinWidthMediaQuery("xs");
```
Comment (qroll) [inline e2e/nextjs-app/src/app/components/theme/media-query.e2e.tsx:17]: include xxs as well (edge case)

File: e2e/tests/components/theme/media-query.e2e.spec.ts [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+            await expect(story.locators.orientationPortraitStatus).toHaveText(
+                "false"
+            );
+        });
+
+        test("portrait orientation detection", async ({ story }) => {
+            await story.init("media-query", {
+                size: "mobile",
+                orientation: "portrait",
+            });
```
Comment (qroll) [inline e2e/tests/components/theme/media-query.e2e.spec.ts]: I think the same `size` should be used here to ensure the values actually change

**Review by qroll — CHANGES_REQUESTED** [review]

File: src/theme/utils/use-resolve-breakpoint-token.ts (hunk 3 above)
Comment (ghazwan-gt) [reply src/theme/utils/use-resolve-breakpoint-token.ts]: Well, we can do that, but doing so will restrict the value to be a css variable name. Now that I think about it more, using the library means using the defined breakpoint when accessing the theme. Will update this

Comment (qroll) [reply src/theme/utils/use-resolve-breakpoint-token.ts]: I was under the impression that `useResolvedTokenValue` still lets you pass in a non-token value

File: src/theme/tokens/media-query.ts
Comment (qroll) [reply src/theme/tokens/media-query.ts:1]: I see that `useOrientation` was removed but this is still here

I thought it would be the other way round, since we intend to do runtime detection for esignature

File: e2e/tests/components/theme/media-query.e2e.spec.ts (hunk 7 above)
Comment (qroll) [reply e2e/tests/components/theme/media-query.e2e.spec.ts]: resolving for now since `useOrientation` was removed

File: src/theme/utils/use-resolve-breakpoint-token.ts (hunk 3 above)
Comment (qroll) [reply src/theme/utils/use-resolve-breakpoint-token.ts]: on this note, instead of `BreakpointCSSVariableString` we should also limit to the `<size>-min` and `<size-max`

because the column/gutter/margins are not relevant to media query

File: src/theme/tokens/media-query.ts
Comment (ghazwan-gt) [reply src/theme/tokens/media-query.ts:1]: I don't think`useOrientation` still has it use since it will be used only by esignature, hence the removal. It can be handled in the component itself later.

I missed deleting the one in this file, will remove

Comment (qroll) [reply src/theme/tokens/media-query.ts:1]: it feels awkward for the component to have to register its own matchMedia listener when there's already a hook. I would suggest to extend `useMediaQuery` to support custom clauses in that case then, when working on the esignature component

**Comment by qroll** [comment]
Layout ColDiv lg test is failing due to the change in orientation. should we revert the changes to the e2e utils as well?

File: src/theme/utils/use-resolve-breakpoint-token.ts (hunk 3 above)
Comment (ghazwan-gt) [reply src/theme/utils/use-resolve-breakpoint-token.ts]: Ah right, I forgot that it include those as well. Will be updating the type

**Review by qroll — APPROVED** [review]
